ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,476 @@
+    /**
+     * DeleteBlob with the coordinator-backed router.
+     */
+    DELETE_COORD(false, false),
+    /**
+     * GetBlobInfo with the coordinator-backed router. Expect an exception to occur because the blob should have
+     * already been deleted.
+     */
+    GET_INFO_DELETED_COORD(false, true),
+    /**
+     * GetBlob with the coordinator-backed router. Expect an exception to occur because the blob should have already
+     * been deleted.
+     */
+    GET_DELETED_COORD(false, true),
+    /**
+     * Wait for the operation chain's blob ID to be reported as created on all replicas. Continue with the remaining
+     * actions in the operation chain afterwards.
+     */
+    AWAIT_CREATION(false, false),
+    /**
+     * Wait for the operation chain's blob ID to be reported as deleted on all replicas. Continue with the remaining
+     * actions in the operation chain afterwards.
+     */
+    AWAIT_DELETION(false, false);
+
+    final boolean nonBlockingRouter;
+    final boolean afterDelete;
+
+    OperationType(boolean nonBlockingRouter, boolean afterDelete) {
+      this.nonBlockingRouter = nonBlockingRouter;
+      this.afterDelete = afterDelete;
+    }
+  }
+
+  /**
+   * Describes an operation chain and provides useful metadata and operation results for testing.
+   */
+  static class OperationInfo {
+    final int operationId;
+    final BlobProperties properties;
+    final byte[] userMetadata;
+    final byte[] data;
+    final Queue<OperationType> opChain;
+    final List<TestFuture> futures = new ArrayList<>();
+    final CountDownLatch latch = new CountDownLatch(1);
+    String blobId;
+
+    OperationInfo(int operationId, BlobProperties properties, byte[] userMetadata, byte[] data,
+        Queue<OperationType> opChain) {
+      this.operationId = operationId;
+      this.properties = properties;
+      this.userMetadata = userMetadata;
+      this.data = data;
+      this.opChain = opChain;
+    }
+  }
+
+  /**
+   * This class encapsulates a future and allows the user to define a check method that runs tests on
+   * the retrieved value in the future.
+   *
+   * @param <T> The type of the encapsulated future
+   */
+  private static abstract class TestFuture<T> {
+    final Future<T> future;
+    final String operationType;
+    final OperationInfo opInfo;
+
+    TestFuture(Future<T> future, String operationType, OperationInfo opInfo) {
+      this.future = future;
+      this.operationType = operationType;
+      this.opInfo = opInfo;
+    }
+
+    /**
+     * Generate a name for the tested operation
+     * @return the operation name
+     */
+    String getOperationName() {
+      return operationType + "-" + opInfo.operationId;
+    }
+
+    /**
+     * Return the value inside the future or throw an {@link AssertionError} if an exception
+     * occurred.
+     * @return the value inside the future
+     */
+    T get() {
+      try {
+        return this.future.get();
+      } catch (Exception e) {
```

Comment:
I agree that it is generally good practice.  However, here I wanted to provide the operation name (operation type, chain ID) in the assertion message for debugging purposes.  What do you think?